core/domain/state_domain.py

```diff
@@ -4871,9 +4871,14 @@ def update_interaction_customization_args(
         Raises:
             Exception. The customization arguments are not unique.
         """
-        # Here we use assert so that the customization_arg_dict will become
-        # Dict type at the entry level so that we can use all the methods of
-        # Dict without MyPy complain.
+        # For argument 'customization_args_dict' we have used Mapping type
+        # because we want to allow 'update_interaction_customization_args'
+        # method to accept different subtypes of customization_arg dictionaries,
+        # but the problem with Mapping is that the Mapping does not allow to
+        # update(or set) values because Mapping is a read-only type. To overcome
+        # this issue, we narrowed down the type from Mapping to Dict by using
+        # assert so that while updating or setting a new value MyPy will not
+        # throw any error.
```

Comment:
While I'm ok with using mapping/dict/etc., could you please make the dict cast more specific in terms of what's actually in it? Currently you've cast it very generally to any sort of dict.

Reply:
Done, instead of assertions now cast is used.